### ISSUE TITLE
[Issue] CVE-2023-23397.ps1 'Unable to create the Azure application password'

### ISSUE THREAD

**Issue by nurminens** [microsoft/CSS-Exchange#1655]
Latest script is unable to create Azure application password even when running script with GA- or AA-account.

```
PS C:\TEMP\CVE> Get-EXOMailbox -ResultSize Unlimited | .\CVE-2023-23397.ps1 -Environment "Online" -UseSearchFolders
CVE-2023-23397 script version 23.04.05.1708
Trying to find Microsoft.Exchange.WebServices.dll in the script folder
Microsoft.Exchange.WebServices.dll was found in the script folder
Invoke-WebRequest : The remote server returned an error: (400) Bad Request.
At C:\TEMP\CVE\CVE-2023-23397.ps1:515 char:5
+     Invoke-WebRequest @params
+     ~~~~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : InvalidOperation: (System.Net.HttpWebRequest:HttpWebRequest) [Invoke-WebRequest], WebExc
   eption
    + FullyQualifiedErrorId : WebCmdletWebResponseException,Microsoft.PowerShell.Commands.InvokeWebRequestCommand

Unable to create the Azure application password
```

Tried to delete previous application, but still same error.

**Comment by lusassl-msft** [comment]
@nurminens can you please try the following and let me know if this works with the current version of the script?

1. Delete the application by running: `.\CVE-2023-23397.ps1 -DeleteAzureApplication`
2. Re-create the application by running: `.\CVE-2023-23397.ps1 -CreateAzureApplication`

Does it work or does the script hit a similar error as described in this issue?

**Comment by nurminens** [comment]
Already updated the script, can't remember what errors I got, when tried with previous version of the script.
I have been deleting app and re-creating using two different accounts (GA and AA), but whatever account I use, I get same error.

Step-by-step when already deleted old application; (Also hard-deleted from deleted apps)
```
PS C:\TEMP\CVE> .\CVE-2023-23397.ps1 -CreateAzureApplication
CVE-2023-23397 script version 23.04.05.1708

Did you know?
The new asynchronous search feature is now general available (GA)!
We recommend using it because the search performance is significantly increased.
You can enable it by using the the following parameters:

UseSearchFolders: Enable deep-traversal SearchFolders search to significantly improve performance
SearchFolderCleanup: Clean up any SearchFolders left behind by the -UseSearchFolders switch

This version of the script also supports Certificate Based Authentication (CBA)

More information: https://aka.ms/CVE-2023-23397ScriptDocFAQ

User: user@domain.com is already an owner of application: CVE-2023-23397Application
Application: CVE-2023-23397Application created with required permissions. Client ID: _CLIENTID_
#
PS C:\TEMP\CVE> Get-EXOMailbox -ResultSize Unlimited | .\CVE-2023-23397.ps1 -Environment "Online" -UseSearchFolders
CVE-2023-23397 script version 23.04.05.1708
Trying to find Microsoft.Exchange.WebServices.dll in the script folder
Microsoft.Exchange.WebServices.dll was found in the script folder
Invoke-WebRequest : The remote server returned an error: (400) Bad Request.
At C:\TEMP\CVE\CVE-2023-23397.ps1:515 char:5
+     Invoke-WebRequest @params
+     ~~~~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : InvalidOperation: (System.Net.HttpWebRequest:HttpWebRequest) [Invoke-WebRequest], WebException
    + FullyQualifiedErrorId : WebCmdletWebResponseException,Microsoft.PowerShell.Commands.InvokeWebRequestCommand

Unable to create the Azure application password
```

When running audit script, it prompts to browsers and asks me to sign-in to that specific tenant with application owner account and after that I'll get that last error.

In debug txt file;
[04/12/2023 13:21:48] : Calling Invoke-GraphApiRequest
[04/12/2023 13:21:48] : Graph API request failed - no response
[04/12/2023 13:21:48] : Unable to create the Azure application password

**Comment by lusassl-msft** [comment]
Thanks, can you please try to run the script from a new PS session (without the Exchange Online Management module loaded) and run it like this (it's enough to provide just one mailbox as this is only for script troubleshooting):

`.\CVE-2023-23397.ps1 -Environment "Online" -UserMailboxes "user1@contoso.com" -UseSearchFolders`

Does it also fail with?

**Comment by nurminens** [comment]
No matter if running script without EOM -module - same error.
No matter which account I use - same error.

**Comment by nurminens** [comment]
Update on this:
Created cert for CBA and script started like there was never any problems.
There might be problem in our systems / proxy which block such scripts. Don't really know, doesn't have that kind of knowledge nor time to investigate any further. 

Script is now creating search folders to users so we will see, what happens.

**Comment by lusassl-msft** [comment]
@nurminens yes, I expected this to work as the "problematic" call to generate a secret for the app will not be performed when doing CBA.
One more thing that you could double-check: Are there already any app secrets assigned to the application? I would not expect any as the app was newly created as you said before, but please have a look and let us know.

Thanks in advance.

**Comment by nurminens** [comment]
There isn't any secrets in app besides my own cert. And running script with AA -account, so that was never the problem.

**Comment by HuuhaaPete** [comment]
Have the exact same issue.
Started getting this error after last script update, until then everything worked fine.
Recreated the application, no difference.

Get-EXOMailbox -ResultSize Unlimited | .\CVE-2023-23397.ps1 -Environment "Online" -UseSearchFolders
CVE-2023-23397 script version 23.04.05.1708
Trying to find Microsoft.Exchange.WebServices.dll in the script folder
Microsoft.Exchange.WebServices.dll was found in the script folder
Invoke-WebRequest : The remote server returned an error: (400) Bad Request.
At C:\Temp\CVE-2023-23397.ps1:515 char:5
+     Invoke-WebRequest @params
+     ~~~~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : InvalidOperation: (System.Net.HttpWebRequest:HttpWebRequest) [Invoke-WebRequest], WebExc
   eption
    + FullyQualifiedErrorId : WebCmdletWebResponseException,Microsoft.PowerShell.Commands.InvokeWebRequestCommand

Unable to create the Azure application password

**Comment by nurminens** [comment]
Try to create new self-signed certificate (CBA) https://microsoft.github.io/CSS-Exchange/Security/CVE-2023-23397/FAQ/
and run
`Get-EXOMailbox -ResultSize unlimited | .\CVE-2023-23397.ps1 -Environment "Online" -UseSearchFolders -CertificateThumbprint <CertThumbprint>`

**Comment by HuuhaaPete** [comment]
> Try to create new self-signed certificate (CBA) https://microsoft.github.io/CSS-Exchange/Security/CVE-2023-23397/FAQ/ and run `Get-EXOMailbox -ResultSize unlimited | .\CVE-2023-23397.ps1 -Environment "Online" -UseSearchFolders -CertificateThumbprint <CertThumbprint>`

Thanks!
That work a round did the trick!

**Comment by lusassl-msft** [comment]
@HuuhaaPete can you please answer these questions:
- Do you use a proxy server on the machine where the script was executed and failed to create the new application secret?
- Are you running the script by using an account who is a Global Administrator or Application Administrator?

**Comment by HuuhaaPete** [comment]
> @HuuhaaPete can you please answer these questions:
> 
> * Do you use a proxy server on the machine where the script was executed and failed to create the new application secret?
> * Are you running the script by using an account who is a Global Administrator or Application Administrator?

It ran successfully and unsuccessfully from home and company LAN. No proxy, only NAT.
I'm running with Global Admin.

**Comment by lusassl-msft** [comment]
@HuuhaaPete we've just released an updated version (`v23.04.21.1504`) of the script. It would be great if you could test it and let us know if the `(400) Bad Request` error when creating the application secrets.

**Comment by HuuhaaPete** [comment]
> @HuuhaaPete we've just released an updated version (`v23.04.21.1504`) of the script. It would be great if you could test it and let us know if the `(400) Bad Request` error still occurs when creating the application secrets.

Script version 23.04.05.170 wouldn't auto update, so tried it manually:
_CVE-2023-23397.ps1 -ScriptUpdateOnly
CVE-2023-23397 script version 23.04.05.1708
No update of the script performed_

Downloaded it manually and now it seems to work on the LAN server with no workaround attempted.
